Move Render command as a CNAB custom action

silvin-lubecki opened the pull request:
**- What I did**
I refactored and cleaned some things in the produced `bundle.json`, including:
* using reverse DNS for custom parameters
* add `Apply-To` action indications for custom parameters
* add `stateless` new indication for custom actions

Then I introduced a new `render` custom action, adding also a new custom parameter to select `yaml|json` output format, and moved the render code into the `cnab-run` binary.

**- How to verify it**

`render` should still work!
```console
./bin/docker-app-standalone render examples/hello-world/hello-world.dockerapp
version: "3.6"
services:
  hello:
    command:
    - -text
    - Hello, World!
    image: hashicorp/http-echo
    ports:
    - mode: ingress
      target: 5678
      published: 8080
      protocol: tcp
```

**- Description for the changelog**
* Transform render command as a CNAB custom action
* Inspect is now a stateless custom action
* Add Apply-To actions indication to custom parameters
* Change custom parameters prefix from "docker" to reverse DNS "com.docker.app"

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/31478878/54626769-25a7ec80-4a72-11e9-9ca5-343cf0a61480.png)

ijc commented on internal/packager/cnab.go:
```diff
@@ -81,7 +81,8 @@ func ToCNAB(app *types.App, invocationImageName string) (*bundle.Bundle, error)
 			internal.Namespace + "inspect": {
-				Modifies: false,
+				Modifies:  false,
+				Stateless: true,
```
I suppose/hope this means you can now drop the dummy creds from `runInspect`?

silvin-lubecki replied:
Right! I forgot this one, thanks for spotting it!

silvin-lubecki replied:
Fixed

ijc commented on internal/commands/render.go:
```diff
@@ -1,74 +1,88 @@
+		defer f.Close()
+		writer = f
+	}
+	err = a.Run(c, map[string]string{"docker.context": ""}, writer)
```
Can these dummy creds not be dropped?

silvin-lubecki replied:
Fixed

silvin-lubecki commented:
Happy to let you handle the rebase/conflicts :trollface:

silvin-lubecki commented:
PTAL @ijc as I cleaned lots of things since your last review:
* bumped duffle with more cleaning in the model (lots of `omitempty` missing) and in the e2e tests
* added a way to specify which stdout/stderr we want while running the invocation image

ijc commented on e2e/commands_test.go:
```diff
@@ -18,7 +18,13 @@ import (
+// the DOCKERAPP_RENDERERS environment variable inside the invocation image,
+// and the only way to do it is to add a new parameter to the bundle.json.
+// This test should be unskipped or removed as soon as the templates are
+// not experimental anymore, or are removed.
```
In principal couldn't we solve that now, by having code which adds that parameter hardcoded when building in experimental mode (and requiring matching experimental enablement in the tool and the invoc image base, which I think is reasonable)?

To be clear though, I can for sure see an argument for not putting a lot of effort into experimental mode right now since we aren't really actively using it.

silvin-lubecki replied:
To be honest I don't wan't to spend more time than I already have on that. I think we can fill another issue on that subject and work on it after the release.

ijc replied:
Sure, I was really just saying that the comment is a spurious justification when there is an actual justification..

ijc commented on e2e/commands_test.go:
```diff
@@ -58,8 +66,14 @@ func testRenderApp(appPath string, env ...string) func(*testing.T) {
+		cmd.Command = append(cmd.Command, "--output="+dir.Join("actual.yaml"))
+		result = icmd.RunCmd(cmd).Assert(t, icmd.Success)
+
+		assert.Assert(t, is.Equal(readFile(t, filepath.Join(appPath, "expected.txt")), readFile(t, dir.Join("actual.yaml"))), "rendering mismatch")
```
nit: that blank line seem spurious...

this looks like an opportunity to use `golden` but since you are following the existing idiom in this file lets not worry about that.

silvin-lubecki replied:
I tried then reverted an attempt to use `golden`, but there is a trick with the filepath, as golden append itself the directory `testdata` to the path, and the resulting code was really ugly...

ijc replied:
Ah, I see that the path to `expected.txt` is a bit dynamic. That makes sense then. Thanks.

ijc commented on e2e/testdata/attachments.dockerapp/docker-compose.yml:
```diff
@@ -16,11 +16,11 @@ services:
       resources:
         limits:
-          memory: $memory
+          memory: ${memory}
```
Any idea what changed/how this was passing before?

silvin-lubecki replied:
I honestly don't know... It's totally weird...

ijc replied:
OK then ;-)

ijc commented on e2e/testdata/simple-bundle.json.golden:
```diff
@@ -5,118 +5,125 @@
 			"name": "John Developer",
-			"email": "john.dev@acmecorp.cool",
-			"url": ""
+			"email": "john.dev@acmecorp.cool"
```
super pedantic: `cool` is (apparently) a real TLD. `acmecorp` is not currently registered there but there's no real reason it couldn't be (then pity poor `john.dev` ;-)).

TL;DR: We should use [RFC2606](https://tools.ietf.org/html/rfc2606) domains for any place holder domain name.

silvin-lubecki replied:
Is `domain.test` ok 👍 ?

ijc replied:
Section 2 of https://tools.ietf.org/html/rfc2606 says so, yes. The other common one to use is `example.{net,org,com}` which are reserved in section 3.

ijc commented on e2e/testdata/simple-bundle.json.golden:
```diff
@@ -5,118 +5,125 @@
 	"parameters": {
 		"api_host": {
 			"type": "string",
 			"defaultValue": "coolapp.com",
```
pendantry: this actually exists (cf RFC2606 comment above)

silvin-lubecki commented:
@ijc PTAL

ijc commented on Gopkg.toml:
```diff
@@ -34,15 +34,17 @@ required = ["github.com/wadey/gocovmerge"]
   name = "github.com/containerd/containerd"
   source = "github.com/simonferquel/containerd"
-  revision = "42c3614d58767e96c1752977fed94d0b140a0173"
+  branch = "push-can-mount"
```
Simon has pushed the commit back, you should switch back to the commit (else changes to the branch will break things).

silvin-lubecki replied:
Fixed

silvin-lubecki commented:
Got rid of the `ride` 😄
